File: AliyunOSSiOSTests/SequentialMultipartUploadTests.m
//
//  SequentialMultipartUploadTests.m
//  AliyunOSSiOSTests
//
//  Created by huaixu on 2018/1/18.
//  Copyright © 2018年 aliyun. All rights reserved.
//

#import "AliyunOSSTests.h"

@interface SequentialMultipartUploadTests : AliyunOSSTests

@end

@implementation SequentialMultipartUploadTests

- (void)setUp {
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testAPI_sequentialMultipartUpload_crcClosed {
    // This is an example of a functional test case.
    // Use XCTAssert and related functions to verify your tests produce the correct results.
    OSSResumableUploadRequest *request = [OSSResumableUploadRequest new];
    request.bucketName = OSS_BUCKET_PUBLIC;
    request.objectKey = @"sequential-multipart";
    request.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    request.deleteUploadIdOnCancelling = NO;
    request.crcFlag = OSSRequestCRCClosed;
//    request.contentSHA1 = [OSSUtil sha1WithFilePath:request.uploadingFileURL.path];
    
    OSSTask *task = [self.client sequentialMultipartUpload:request];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNil(t.error);
        
        return nil;
    }] waitUntilFinished];
}

- (void)testAPI_sequentialMultipartUpload_crcOpen {
    // This is an example of a functional test case.
    // Use XCTAssert and related functions to verify your tests produce the correct results.
    OSSResumableUploadRequest *request = [OSSResumableUploadRequest new];
    request.bucketName = OSS_BUCKET_PUBLIC;
    request.objectKey = @"sequential-multipart";
    request.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    request.deleteUploadIdOnCancelling = NO;
    request.crcFlag = OSSRequestCRCOpen;
//    request.contentSHA1 = [OSSUtil sha1WithFilePath:request.uploadingFileURL.path];
    
    OSSTask *task = [self.client sequentialMultipartUpload:request];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNil(t.error);
        
        return nil;
    }] waitUntilFinished];
}

- (void)testAPI_sequentialMultipartUpload_cancel_withoutDeleteRecord {
    OSSResumableUploadRequest *request = [OSSResumableUploadRequest new];
    request.bucketName = OSS_BUCKET_PUBLIC;
    request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
    request.objectKey = @"sequential-multipart";
    request.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    request.deleteUploadIdOnCancelling = NO;
    request.crcFlag = OSSRequestCRCOpen;
    __weak typeof(request) weakRequest = request;
    request.uploadProgress = ^(int64_t bytesSent, int64_t totalBytesSent, int64_t totalBytesExpectedToSend) {
        if (totalBytesSent > totalBytesExpectedToSend / 2) {
            [weakRequest cancel];
        }
    };
    
    OSSTask *task = [self.client sequentialMultipartUpload:request];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNotNil(t.error);
        XCTAssertEqual(t.error.code, OSSClientErrorCodeTaskCancelled);
        
        return nil;
    }] waitUntilFinished];
}

- (void)testAPI_sequentialMultipartUpload_cancel_deleteRecord {
    OSSResumableUploadRequest *request = [OSSResumableUploadRequest new];
    request.bucketName = OSS_BUCKET_PUBLIC;
    request.objectKey = @"sequential-multipart";
    request.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
    request.deleteUploadIdOnCancelling = YES;
    request.crcFlag = OSSRequestCRCOpen;
    __weak typeof(request) weakRequest = request;
    request.uploadProgress = ^(int64_t bytesSent, int64_t totalBytesSent, int64_t totalBytesExpectedToSend) {
        if (totalBytesSent > totalBytesExpectedToSend / 2) {
            [weakRequest cancel];
        }
    };
    
    OSSTask *task = [self.client sequentialMultipartUpload:request];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNotNil(t.error);
        XCTAssertEqual(t.error.code, OSSClientErrorCodeTaskCancelled);
        
        return nil;
    }] waitUntilFinished];
}

- (void)testAPI_sequentialMultipartUpload_cancel_and_resume_crcClosed {
    OSSResumableUploadRequest *request = [OSSResumableUploadRequest new];
    request.bucketName = OSS_BUCKET_PUBLIC;
    request.objectKey = @"sequential-multipart";
    request.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    request.deleteUploadIdOnCancelling = NO;
    request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
    request.crcFlag = OSSRequestCRCClosed;
    __weak typeof(request) weakRequest = request;
    request.uploadProgress = ^(int64_t bytesSent, int64_t totalBytesSent, int64_t totalBytesExpectedToSend) {
        if (totalBytesSent > totalBytesExpectedToSend / 2) {
            [weakRequest cancel];
        }
    };
    
    OSSTask *task = [self.client sequentialMultipartUpload:request];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNotNil(t.error);
        XCTAssertEqual(t.error.code, OSSClientErrorCodeTaskCancelled);
        
        return nil;
    }] waitUntilFinished];
    
    OSSResumableUploadRequest *resumedRequest = [OSSResumableUploadRequest new];
    resumedRequest.bucketName = OSS_BUCKET_PUBLIC;
    resumedRequest.objectKey = @"sequential-multipart";
    resumedRequest.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    resumedRequest.deleteUploadIdOnCancelling = NO;
    resumedRequest.crcFlag = OSSRequestCRCClosed;
    resumedRequest.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
//    resumedRequest.contentSHA1 = [OSSUtil sha1WithFilePath:request.uploadingFileURL.path];
    
    task = [self.client sequentialMultipartUpload:resumedRequest];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNil(t.error);
        return nil;
    }] waitUntilFinished];
}

- (void)testAPI_sequentialMultipartUpload_cancel_and_resume_crcOpened {
    OSSResumableUploadRequest *request = [OSSResumableUploadRequest new];
    request.bucketName = OSS_BUCKET_PUBLIC;
    request.objectKey = @"sequential-multipart";
    request.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    request.deleteUploadIdOnCancelling = NO;
    request.crcFlag = OSSRequestCRCOpen;
    request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
    
    __weak typeof(request) weakRequest = request;
    request.uploadProgress = ^(int64_t bytesSent, int64_t totalBytesSent, int64_t totalBytesExpectedToSend) {
        if (totalBytesSent > totalBytesExpectedToSend / 2) {
            [weakRequest cancel];
        }
    };
    
    OSSTask *task = [self.client sequentialMultipartUpload:request];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNotNil(t.error);
        XCTAssertEqual(t.error.code, OSSClientErrorCodeTaskCancelled);
        
        return nil;
    }] waitUntilFinished];
    
    OSSResumableUploadRequest *resumedRequest = [OSSResumableUploadRequest new];
    resumedRequest.bucketName = OSS_BUCKET_PUBLIC;
    resumedRequest.objectKey = @"sequential-multipart";
    resumedRequest.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    resumedRequest.deleteUploadIdOnCancelling = NO;
    resumedRequest.crcFlag = OSSRequestCRCOpen;
    resumedRequest.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
//    resumedRequest.contentSHA1 = [OSSUtil sha1WithFilePath:request.uploadingFileURL.path];
    
    task = [self.client sequentialMultipartUpload:resumedRequest];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNil(t.error);
        
        return nil;
    }] waitUntilFinished];
}

- (void)testAPI_sequentialMultipartUpload_cancel_and_resume_lastCrcOpened {
    OSSResumableUploadRequest *request = [OSSResumableUploadRequest new];
    request.bucketName = OSS_BUCKET_PUBLIC;
    request.objectKey = @"sequential-multipart";
    request.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    request.deleteUploadIdOnCancelling = YES;
    request.crcFlag = OSSRequestCRCClosed;
    request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
    
    __weak typeof(request) weakRequest = request;
    request.uploadProgress = ^(int64_t bytesSent, int64_t totalBytesSent, int64_t totalBytesExpectedToSend) {
        if (totalBytesSent > totalBytesExpectedToSend / 2) {
            [weakRequest cancel];
        }
    };
    
    OSSTask *task = [self.client sequentialMultipartUpload:request];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNotNil(t.error);
        XCTAssertEqual(t.error.code, OSSClientErrorCodeTaskCancelled);
        
        return nil;
    }] waitUntilFinished];
    
    OSSResumableUploadRequest *resumedRequest = [OSSResumableUploadRequest new];
    resumedRequest.bucketName = OSS_BUCKET_PUBLIC;
    resumedRequest.objectKey = @"sequential-multipart";
    resumedRequest.uploadingFileURL = [[NSBundle mainBundle] URLForResource:@"wangwang" withExtension:@"zip"];
    resumedRequest.deleteUploadIdOnCancelling = NO;
    resumedRequest.crcFlag = OSSRequestCRCOpen;
    resumedRequest.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject;
//    resumedRequest.contentSHA1 = [OSSUtil sha1WithFilePath:request.uploadingFileURL.path];
    
    task = [self.client sequentialMultipartUpload:resumedRequest];
    [[task continueWithBlock:^OSSTask* (OSSTask* t) {
        XCTAssertNil(t.error);
        
        return nil;
    }] waitUntilFinished];
}

@end

File: OSSSwiftDemo/OSSSwiftDemoTests/SequentialMultipartUploadTests.swift
//
//  SequentialMultipartUploadTests.swift
//  OSSSwiftDemoTests
//
//  Created by huaixu on 2018/1/18.
//  Copyright © 2018年 aliyun. All rights reserved.
//

import XCTest
import AliyunOSSiOS
import AliyunOSSSwiftSDK

class SequentialMultipartUploadTests: OSSSwiftDemoTests {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    override func setupClient() {
        let provider = OSSPlainTextAKSKPairCredentialProvider.init(plainTextAccessKey: OSS_ACCESSKEY_ID, secretKey: OSS_SECRETKEY_ID)
        client = OSSClient.init(endpoint: OSS_ENDPOINT, credentialProvider: provider)
    }
    
    func testAPI_testAPI_sequentialMultipartUpload_crcClosed() {
        let request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = false
        request.crcFlag = OSSRequestCRCFlag.open
        let filePath = Bundle.main.path(forResource: "wangwang", ofType: "zip")
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        
        let task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNil(t.error)
            
            return nil;
        }).waitUntilFinished()
    }
    
    func testAPI_sequentialMultipartUpload_crcOpen() {
        let request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = false
        request.crcFlag = OSSRequestCRCFlag.closed
        let filePath = Bundle.main.path(forResource: "wangwang", ofType: "zip")
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        
        let task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNil(t.error);
            
            return nil
        }).waitUntilFinished()
    }
    
    func testAPI_sequentialMultipartUpload_cancel_withoutDeleteRecord() {
        let request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = false
        request.crcFlag = OSSRequestCRCFlag.open
        request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        request.uploadProgress = {[weak request](bytesSent, totalBytesSent, totalBytesExpectedToSend) in
            if totalBytesSent > totalBytesExpectedToSend / 2 {
                request?.cancel()
            }
        }
        
        let filePath = Bundle.main.path(forResource: "wangwang", ofType: "zip")
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        
        let task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNotNil(t.error)
            let error = t.error! as NSError
            XCTAssertEqual(error.code, OSSClientErrorCODE.codeTaskCancelled.rawValue)
            
            return nil
        }).waitUntilFinished()
    }
    
    func testAPI_sequentialMultipartUpload_cancel_deleteRecord() {
        let request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = true
        request.crcFlag = OSSRequestCRCFlag.open
        request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        request.uploadProgress = {[weak request](bytesSent, totalBytesSent, totalBytesExpectedToSend) in
            if totalBytesSent > totalBytesExpectedToSend / 2 {
                request?.cancel()
            }
        }
        
        let filePath = Bundle.main.path(forResource: "wangwang", ofType: "zip")
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        
        let task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNotNil(t.error);
            let error = t.error! as NSError
            XCTAssertEqual(error.code, OSSClientErrorCODE.codeTaskCancelled.rawValue)
            
            return nil;
        }).waitUntilFinished()
    }
    
    func testAPI_sequentialMultipartUpload_cancel_and_resume_crcClosed() {
        var request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = false
        request.crcFlag = OSSRequestCRCFlag.closed
        request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        request.uploadProgress = {[weak request](bytesSent, totalBytesSent, totalBytesExpectedToSend) in
            if totalBytesSent > totalBytesExpectedToSend / 2 {
                request?.cancel()
            }
        }
        
        let filePath = Bundle.main.path(forResource: "wangwang", ofType: "zip")
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        
        var task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNotNil(t.error)
            let error = t.error! as NSError
            XCTAssertEqual(error.code, OSSClientErrorCODE.codeTaskCancelled.rawValue)
            
            return nil;
        }).waitUntilFinished()
        
        request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = true
        request.crcFlag = OSSRequestCRCFlag.closed
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        request.uploadProgress = {(bytesSent, totalBytesSent, totalBytesExpectedToSend) in
            print("bytesSent: \(bytesSent), totalBytesSent: \(totalBytesSent), totalBytesExpectedToSend: \(totalBytesExpectedToSend)")
        }
        
        task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNil(t.error)
            
            return nil
        }).waitUntilFinished()
    }
    
    func testAPI_sequentialMultipartUpload_cancel_and_resume_crcOpened() {
        var request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = false
        request.crcFlag = OSSRequestCRCFlag.open
        request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        request.uploadProgress = {[weak request](bytesSent, totalBytesSent, totalBytesExpectedToSend) in
            if totalBytesSent > totalBytesExpectedToSend / 2 {
                request?.cancel()
            }
        }
        
        let filePath = Bundle.main.path(forResource: "wangwang", ofType: "zip")
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        
        var task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNotNil(t.error)
            let error = t.error! as NSError
            XCTAssertEqual(error.code, OSSClientErrorCODE.codeTaskCancelled.rawValue)
            
            return nil;
        }).waitUntilFinished()
        
        request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = true
        request.crcFlag = OSSRequestCRCFlag.open
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        request.uploadProgress = {(bytesSent, totalBytesSent, totalBytesExpectedToSend) in
            print("bytesSent: \(bytesSent), totalBytesSent: \(totalBytesSent), totalBytesExpectedToSend: \(totalBytesExpectedToSend)")
        }
        
        task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNil(t.error)
            
            return nil
        }).waitUntilFinished()
    }
    
    func testAPI_sequentialMultipartUpload_cancel_and_resume_lastCrcOpened() {
        var request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = false
        request.crcFlag = OSSRequestCRCFlag.closed
        request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        request.uploadProgress = {[weak request](bytesSent, totalBytesSent, totalBytesExpectedToSend) in
            if totalBytesSent > totalBytesExpectedToSend / 2 {
                request?.cancel()
            }
        }
        
        let filePath = Bundle.main.path(forResource: "wangwang", ofType: "zip")
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        
        var task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNotNil(t.error)
            let error = t.error! as NSError
            XCTAssertEqual(error.code, OSSClientErrorCODE.codeTaskCancelled.rawValue)
            
            return nil;
        }).waitUntilFinished()
        
        request = OSSResumableUploadRequest()
        request.bucketName = OSS_BUCKET_PUBLIC;
        request.objectKey = "sequential-swift-multipart";
        request.uploadingFileURL = Bundle.main.url(forResource: "wangwang", withExtension: "zip")!
        request.deleteUploadIdOnCancelling = true
        request.crcFlag = OSSRequestCRCFlag.open
        request.contentSHA1 = OSSUtil.sha1(withFilePath: filePath)
        request.recordDirectoryPath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!
        request.uploadProgress = {(bytesSent, totalBytesSent, totalBytesExpectedToSend) in
            print("bytesSent: \(bytesSent), totalBytesSent: \(totalBytesSent), totalBytesExpectedToSend: \(totalBytesExpectedToSend)")
        }
        
        task = client.sequentialMultipartUpload(request)
        task.continue({ (t) -> Any? in
            XCTAssertNil(t.error)
            
            return nil
        }).waitUntilFinished()
    }
    
}
